### PR TITLE
Step3

### PR DESCRIPTION
**Pull request by legendado**

### CODE REVIEW

File: spec/features/catalog_spec.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  it 'click on View More button' do
+    click_link(view_more)
+    expect(page).to have_current_path(books_path)
+    expect(page).to have_content(Book.order(title: :asc).last.title)
+    expect(page).to have_no_link(view_more)
+  end
+
+  it 'redirect to book page' do
```
Comment (eugene1599) [inline spec/features/catalog_spec.rb]: it's not clear what do you try to test here

File: spec/features/catalog_spec.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  it 'sort books' do
+    click_link(I18n.t('catalog.sortings.title_asc'), match: :first)
+    click_link(I18n.t('catalog.sortings.newest'), match: :first)
+    expect(find(book_selector, match: :first)).to have_content(Book.last.title)
+  end
+
+  it 'click on View More button' do
+    click_link(view_more)
+    expect(page).to have_current_path(books_path)
+    expect(page).to have_content(Book.order(title: :asc).last.title)
```
Comment (eugene1599) [inline spec/features/catalog_spec.rb]: you can create 2 books for this test, stub constant with number of books on the page

File: spec/features/catalog_spec.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  let!(:books_from_first_category) { create_list(:book, 5, category: create(:category)) }
+  let!(:books_from_second_category) { create_list(:book, 5, category: create(:category)) }
+
+  before { visit(books_path) }
+
+  it 'choose category' do
+    find('.filter-link', text: first_category.name.capitalize, match: :first).click
+    expect(page).to have_current_path(category_books_path(first_category))
+    expect(page).to have_selector(book_selector, count: books_from_first_category.count)
+    expect(page).to have_content(books_from_first_category.first.title)
```
Comment (eugene1599) [inline spec/features/catalog_spec.rb]: what about other 4 books?

File: spec/features/catalog_spec.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    expect(page).to have_no_content(books_from_second_category.first.title)
```
Comment (eugene1599) [inline spec/features/catalog_spec.rb]: what about other 4 books?

File: spec/features/catalog_spec.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
+describe 'Catalog page', type: :feature, js: true do
+  let(:first_category) { Category.first }
```
Comment (eugene1599) [inline spec/features/catalog_spec.rb]: create(:category) and use this category for created books?

File: spec/features/home_spec.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
+describe 'Home page', type: :feature, js: true do
+  let(:category) { Category.first }
+  let(:latest_book_selector) { '.item' }
+  let(:shop) { I18n.t('navbar.shop') }
+  let(:books) { I18n.t('navbar.books') }
+  let(:get_started) { I18n.t('buttons.get_started') }
+  let(:latest_book_count) { 3 }
```
Comment (eugene1599) [inline spec/features/home_spec.rb]: use value from constant

File: spec/features/home_spec.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
+describe 'Home page', type: :feature, js: true do
+  let(:category) { Category.first }
```
Comment (eugene1599) [inline spec/features/home_spec.rb]: why do you need this?

File: spec/features/home_spec.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    end
+
+    it 'choose Books from footer' do
+      within 'footer' do
+        click_link(books)
+        expect(page).to have_current_path(books_path)
+      end
+    end
+
+    it 'click on Get Started button' do
```
Comment (eugene1599) [inline spec/features/home_spec.rb]: do you check `Get Started` button presence here?

File: spec/features/catalog_spec.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+  it 'choose category' do
+    find('.filter-link', text: first_category.name.capitalize, match: :first).click
+    expect(page).to have_current_path(category_books_path(first_category))
+    expect(page).to have_selector(book_selector, count: books_from_first_category.count)
+    expect(page).to have_content(books_from_first_category.first.title)
+    expect(page).to have_no_content(books_from_second_category.first.title)
+  end
+
+  it 'sort books' do
```
Comment (eugene1599) [inline spec/features/catalog_spec.rb]: what about other sorting? (by price 0-9, 9-0, title a-z, title z-a and other)

File: app/views/devise/registrations/edit.html.erb [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+<%= form_for(resource, as: resource_name, url: registration_path(resource_name), html: { method: :put }) do |f| %>
+  <%= render "devise/shared/error_messages", resource: resource %>
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true, autocomplete: "email" %>
+  </div>
+
+  <% if devise_mapping.confirmable? && resource.pending_reconfirmation? %>
+    <div>Currently waiting confirmation for: <%= resource.unconfirmed_email %></div>
```
[the remark this answers is not in the document]
Comment (legendado) [reply app/views/devise/registrations/edit.html.erb]: this file is completely changed in the next step

File: app/views/devise/registrations/edit.html.erb [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  <% end %>
+
+  <div class="field">
+    <%= f.label :password %> <i>(leave blank if you don't want to change it)</i><br />
```
[the remark this answers is not in the document]
Comment (legendado) [reply app/views/devise/registrations/edit.html.erb]: this file is completely changed in the next step

File: app/views/devise/registrations/edit.html.erb [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    <% end %>
+  </div>
+
+  <div class="field">
+    <%= f.label :password_confirmation %><br />
+    <%= f.password_field :password_confirmation, autocomplete: "new-password" %>
+  </div>
+
+  <div class="field">
+    <%= f.label :current_password %> <i>(we need your current password to confirm your changes)</i><br />
```
[the remark this answers is not in the document]
Comment (legendado) [reply app/views/devise/registrations/edit.html.erb]: this file is completely changed in the next step

File: app/views/devise/registrations/edit.html.erb [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    <%= f.password_field :current_password, autocomplete: "current-password" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Update" %>
+  </div>
+<% end %>
+
+<h3>Cancel my account</h3>
```
[the remark this answers is not in the document]
Comment (legendado) [reply app/views/devise/registrations/edit.html.erb]: this file is completely changed in the next step

File: app/views/devise/registrations/edit.html.erb [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    <%= f.password_field :password_confirmation, autocomplete: "new-password" %>
+  </div>
+
+  <div class="field">
+    <%= f.label :current_password %> <i>(we need your current password to confirm your changes)</i><br />
+    <%= f.password_field :current_password, autocomplete: "current-password" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Update" %>
```
[the remark this answers is not in the document]
Comment (legendado) [reply app/views/devise/registrations/edit.html.erb]: this file is completely changed in the next step

File: app/views/devise/registrations/edit.html.erb [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+  </div>
+<% end %>
+
+<h3>Cancel my account</h3>
+
+<p>Unhappy? <%= button_to "Cancel my account", registration_path(resource_name), data: { confirm: "Are you sure?" }, method: :delete %></p>
```
[the remark this answers is not in the document]
Comment (legendado) [reply app/views/devise/registrations/edit.html.erb]: this file is completely changed in the next step

File: app/views/devise/registrations/edit.html.erb [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+<%= link_to "Back", :back %>
```
[the remark this answers is not in the document]
Comment (legendado) [reply app/views/devise/registrations/edit.html.erb]: this file is completely changed in the next step

File: app/views/devise/registrations/edit.html.erb [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+  <% if devise_mapping.confirmable? && resource.pending_reconfirmation? %>
+    <div>Currently waiting confirmation for: <%= resource.unconfirmed_email %></div>
+  <% end %>
+
+  <div class="field">
+    <%= f.label :password %> <i>(leave blank if you don't want to change it)</i><br />
+    <%= f.password_field :password, autocomplete: "new-password" %>
+    <% if @minimum_password_length %>
+      <br />
+      <em><%= @minimum_password_length %> characters minimum</em>
```
[the remark this answers is not in the document]
Comment (legendado) [reply app/views/devise/registrations/edit.html.erb]: this file is completely changed in the next step

File: spec/features/authentication/sign_in_spec.rb [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+      click_button(I18n.t('buttons.log_in'))
+    end
+
+    it { expect(page).to have_current_path(root_path) }
+    it { expect(page).to have_content(sucessful_sign_in) }
+  end
+
+  context 'when invalid data' do
+    let(:invalid_email) { 'invalid@email.com' }
+    let(:key) { 'Email' }
```
[the remark this answers is not in the document]
Comment (legendado) [reply spec/features/authentication/sign_in_spec.rb]: i don't have i18n constant for this

File: spec/queries/books_query_spec.rb [hunk 19]
```diff
@@ -15,16 +15,29 @@
...
+
+    let(:params) do
+      lambda { |sort_type|
+        { sort: sort_type }
+      }
+    end
 
     it 'sort by default' do
-      expect(described_class.call).to match_array(books_with_default_sort)
+      expect(described_class.call).to match_array(books_with_sort[sortings[default_sort.to_sym]])
```
Comment (eugene1599) [inline spec/queries/books_query_spec.rb]: please simplify your tests

File: spec/queries/books_query_spec.rb [hunk 20]
```diff
@@ -15,16 +15,29 @@
...
 
   describe '.sort' do
-    let(:params) { { sort: 'price_asc' } }
-    let(:books_with_price_sort) { Book.order(price: :asc) }
-    let(:books_with_default_sort) { Book.order(title: :asc) }
+    let(:default_sort) { BooksQuery::DEFAULT_SORTING }
+    let(:sortings) { BooksQuery::SORTINGS }
+
+    let(:books_with_sort) do
+      lambda { |sort_type|
```
Comment (eugene1599) [inline spec/queries/books_query_spec.rb]: i think you don't need it here

File: spec/queries/books_query_spec.rb [hunk 21]
```diff
@@ -15,16 +15,29 @@
...
 
     it 'sort by default' do
-      expect(described_class.call).to match_array(books_with_default_sort)
+      expect(described_class.call).to match_array(books_with_sort[sortings[default_sort.to_sym]])
     end
 
     it 'sort by sort type' do
-      expect(described_class.call(params)).to match_array(books_with_price_sort)
+      sortings.each do |type, sort|
+        expect(described_class.call(params[type])).to match_array(books_with_sort[sort])
```
Comment (eugene1599) [inline spec/queries/books_query_spec.rb]: please simplify your tests

File: spec/features/home_spec.rb [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+  context 'when User goes to Catalog page' do
+    it 'choose Category from navbar' do
+      within '#desktop' do
+        click_link(shop)
+        click_link(category.name)
+        expect(page).to have_current_path(category_books_path(category))
+      end
+    end
+
+    it 'choose Books from footer' do
```
Comment (eugene1599) [inline spec/features/home_spec.rb]: what do you want to test here?
- you can choose `books link` from footer 
or
- `books link` from footer opens books_path

??

File: spec/queries/books_query_spec.rb [hunk 23]
```diff
@@ -15,16 +15,29 @@
...
+    let(:default_sort) { BooksQuery::DEFAULT_SORTING }
+    let(:sortings) { BooksQuery::SORTINGS }
+
+    let(:books_with_sort) do
+      lambda { |sort_type|
+        Book.order(sort_type)
+      }
+    end
+
+    let(:params) do
```
Comment (eugene1599) [inline spec/queries/books_query_spec.rb]: why do you need it?

File: app/assets/javascripts/alerts.js [hunk 24]
```diff
@@ -0,0 +1,5 @@
+window.setTimeout(function() {
```
Comment (eugene1599) [inline app/assets/javascripts/alerts.js]: why do you need it?

Comment (legendado) [reply app/assets/javascripts/alerts.js]: i did it to hide automaticly alerts and notices

**Pull request merged**